# BÀI TẬP: EDA HOÀN CHỈNH — TITANIC
## Data Profiling → Descriptive Statistics → Define the Question
**Nguồn:** kaggle.com/c/titanic (891 dòng)

```
PHẦN A — DATA PROFILING       : dữ liệu sạch chưa, có gì
PHẦN B — DESCRIPTIVE STATS    : từng biến trông như thế nào
PHẦN C — DEFINE THE QUESTION  : đặt câu hỏi kinh doanh, trả lời bằng code, rút insight
```

## Setup

In [25]:
import pandas as pd, numpy as np, matplotlib.pyplot as plt, seaborn as sns
from pathlib import Path
from scipy import stats

sns.set_style('whitegrid')

csv_path = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

df = pd.read_csv(csv_path)
print('Loaded from:', csv_path)
df.head()

Loaded from: https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


---
# PHẦN A — DATA PROFILING
## A.1. Data size, column names, data types

In [26]:
# TODO
print(f'Số dòng và số cột: {df.shape}')
print(f'Danh sách tên cột: df.columns.tolist()')
print(f'Data types: {df.dtypes}')

Số dòng và số cột: (891, 15)
Danh sách tên cột: df.columns.tolist()
Data types: survived         int64
pclass           int64
sex             object
age            float64
sibsp            int64
parch            int64
fare           float64
embarked        object
class           object
who             object
adult_male        bool
deck            object
embark_town     object
alive           object
alone             bool
dtype: object


## A.2. Missing values & Duplicate data

In [27]:
# TODO
print(f'Total missing values on each column\n{df.isnull().sum()}')
print(f'Total number of rows with duplicate data: {df.duplicated().sum()}')


Total missing values on each column
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64
Total number of rows with duplicate data: 107


## A.3. Invalid values

In [28]:
print(f'number of rows with invalid ages: {(df['age'] < 0).sum()}')
print(f'number of rows with invalid survival datas: {(~df['survived'].isin([1,0])).sum()}')
print(f'number of rows with invalid fare: {(df['fare'] < 0).sum()}')
print(f'number of rows with empty values: \n{df.isnull().sum()}')

number of rows with invalid ages: 0
number of rows with invalid survival datas: 0
number of rows with invalid fare: 0
number of rows with empty values: 
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


## A.4. Create a new column
Tạo cột `family_size` = sibsp + parch + 1.

In [29]:
# TODO
df['family_size'] = df['sibsp'] + df['parch'] + 1
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone,family_size
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False,2
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False,2
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True,1
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False,2
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True,1
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True,1
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False,4
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True,1


---
# PHẦN B — DESCRIPTIVE STATISTICS
## Group 1 — Central Tendency

In [30]:
for col in ['age','fare','sibsp','parch','pclass']:
    mean_, median_, mode_ = df[col].mean(), df[col].median(), df[col].mode()[0]
    print(f'{col}: mean={mean_:.2f}, median={median_:.2f},mode={mode_:.2f}')

print()

print(f'mode sex: {df['sex'].mode()[0]}')
print(f'mode embarked: {df['embarked'].mode()[0]}')
print(f'mode class: {df['class'].mode()[0]}')
print(f'mode who: {df['who'].mode()[0]}')
print(f'mode embark_town: {df['embark_town'].mode()[0]}')

age: mean=29.70, median=28.00,mode=24.00
fare: mean=32.20, median=14.45,mode=8.05
sibsp: mean=0.52, median=0.00,mode=0.00
parch: mean=0.38, median=0.00,mode=0.00
pclass: mean=2.31, median=3.00,mode=3.00

mode sex: male
mode embarked: S
mode class: Third
mode who: man
mode embark_town: Southampton


## Group 2 — Dispersion

In [31]:
# TODO
for col in ['age','fare']:
    c = df[col]
    # range
    range_ = c.max() - c.min()

    #variance
    var_ = c.var()

    #std
    std_ = c.std()

    #IQR
    q1 = c.quantile(0.25)
    q3 = c.quantile(0.75)
    iqr = q3-q1

    #confficient of variation
    cv_ = std_/c.mean()
    print(f'-----{col}-----')
    print(f'Range = {range_:.2f}, variance = {var_:.2f}, standard deviation = {std_:.2f}, IQR = {iqr}, cv = {cv_}')


-----age-----
Range = 79.58, variance = 211.02, standard deviation = 14.53, IQR = 17.875, cv = 0.4891221855465675
-----fare-----
Range = 512.33, variance = 2469.44, standard deviation = 49.69, IQR = 23.0896, cv = 1.5430725278408497


## Group 3 — Location and Shape

In [40]:
# TODO
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone,family_size,fare_group
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False,2,Thấp
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False,2,Cao
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True,1,Trung bình thấp
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False,2,Cao
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True,1,Trung bình thấp
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True,1,Trung bình thấp
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True,1,Trung bình cao
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False,4,Trung bình cao
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True,1,Trung bình cao


---
# PHẦN C — DEFINE THE QUESTION

## Câu hỏi 1: Hạng vé nào có tỷ lệ sống sót cao nhất, chênh lệch bao nhiêu so với hạng thấp nhất?

In [41]:
survival_class = (df.groupby('class')['survived'].mean() * 100).sort_values(ascending=False)
display(survival_class.round(2))

best_class = survival_class.index[0]
best_rate = survival_class.iloc[0]

difference = survival_class.loc['First'] - survival_class.loc['Third']

print(f"=> Hạng có tỷ lệ sống sót cao nhất là: {best_class} ({best_rate:.2f}%)")
print(f"=> Chênh lệch giữa first và third là: {difference:.2f}%")

class
First     62.96
Second    47.28
Third     24.24
Name: survived, dtype: float64

=> Hạng có tỷ lệ sống sót cao nhất là: First (62.96%)
=> Chênh lệch giữa first và third là: 38.73%


## Câu hỏi 2: Giới tính hay hạng vé ảnh hưởng đến sống sót mạnh hơn?

In [34]:
# sex
sex_survival = df.groupby('sex')['survived'].mean() * 100
difference_sex = sex_survival['female'] - sex_survival['male']

#class
pclass_survival = df.groupby('pclass')['survived'].mean() * 100
difference_pclass = pclass_survival[1] - pclass_survival[3]

#use pviot table
combine_analysis = df.pivot_table(index='sex', columns='pclass', values='survived', aggfunc='mean') * 100

print("---SEX--")
display(sex_survival.round(2))
print(f"=> Chênh lệch giữa nam và nữa: {difference_sex:.2f}%\n")

print("---CLASS---")
display(pclass_survival.round(2))
print(f"=> Chênh lệch giữa hạng 1 và hạng 3: {difference_pclass:.2f}%\n")

print("SEX and CLASS")
display(combine_analysis.round(2))

---SEX--


sex
female    74.20
male      18.89
Name: survived, dtype: float64

=> Chênh lệch giữa nam và nữa: 55.31%

---CLASS---


pclass
1    62.96
2    47.28
3    24.24
Name: survived, dtype: float64

=> Chênh lệch giữa hạng 1 và hạng 3: 38.73%

SEX and CLASS


pclass,1,2,3
sex,,,
female,96.81,92.11,50.00
male,36.89,15.74,13.54


Insight:

## Câu hỏi 3: Vé đắt hơn có thực sự sống sót cao hơn không?

In [35]:
fare_stats = df.groupby('survived')['fare'].agg(
    so_luong='count',
    mean='mean',
    nedian='median',
    std='std',
    min='min',
    max='max'
)

display(fare_stats.round(2))

df['fare_group'] = pd.qcut(df['fare'], q=4, labels=['Thấp', 'Trung bình thấp', 'Trung bình cao', 'Cao'])
fare_group_survival = df.groupby('fare_group', observed=False)['survived'].mean() * 100

print("---tỉ lệ soogns sót theo phân khúc giá")
display(fare_group_survival.round(2))

,so_luong,mean,nedian,std,min,max
survived,,,,,,
0,549,22.12,10.5,31.39,0.0,263.00
1,342,48.40,26.0,66.60,0.0,512.33


---tỉ lệ soogns sót theo phân khúc giá


fare_group
Thấp               19.73
Trung bình thấp    30.36
Trung bình cao     45.50
Cao                58.11
Name: survived, dtype: float64

## Câu hỏi 4: Gia đình đông người có ảnh hưởng đến khả năng sống sót không?

In [36]:
so_luong = df.groupby('family_size')['survived'].count()
ty_le = (df.groupby('family_size')['survived'].mean() * 100)

family_stats = pd.DataFrame({
    'so_luong': so_luong,
    'ty_le_song_sot (%)': ty_le
})

display(family_stats.round(2))

,so_luong,ty_le_song_sot (%)
family_size,,
1,537,30.35
2,161,55.28
3,102,57.84
4,29,72.41
5,15,20.00
6,22,13.64
7,12,33.33
8,6,0.00
11,7,0.00


## Câu hỏi 5: Cảng lên tàu (embark_town) nào có tỷ lệ sống sót cao nhất?

In [38]:

so_luong_cang = df.groupby('embark_town')['survived'].count()
ty_le_cang = df.groupby('embark_town')['survived'].mean() * 100

port_survival = pd.DataFrame({
    'so_luong': so_luong_cang,
    'ty_le_song_sot (%)': ty_le_cang
}).sort_values(by='ty_le_song_sot (%)', ascending=False)

display(port_survival.round(2))

best_port = port_survival.index[0]
best_port_rate = port_survival.iloc[0]['ty_le_song_sot (%)']

print(f"=> Cảng có tỷ lệ sống sót cao nhất là: {best_port} ({best_port_rate:.2f}%)")

,so_luong,ty_le_song_sot (%)
embark_town,,
Cherbourg,168,55.36
Queenstown,77,38.96
Southampton,644,33.70


=> Cảng có tỷ lệ sống sót cao nhất là: Cherbourg (55.36%)


## Câu hỏi 6 (Tổng hợp) — Viết insight tổng hợp
Dựa trên Phần A, B, C, viết 4-5 câu insight tổng thể về dữ liệu Titanic.

1. Hạng vé và giá tiền tạo ra sự phân hóa rõ rệt: Hành khách ở Hạng 1 có tỷ lệ sống sót cao nhất (62.96%), cao hơn hẳn hạng 3 (24.24%) nhờ vị trí các phòng ở tầng trên dễ tiếp cận xuồng cứu hộ hơn
2.Quy mô gia đình lý tưởng là từ 2 đến 4 người: Nhóm đi cùng 1–3 người thân có tỷ lệ sống sót cao nhất  trong khi người đi một mình hoặc đi theo gia đình quá đông  gặp nhiều khó khăn hơn khi tìm kiếm và thoát hiểm
3. Cơ hội sống sót phụ thuộc lớn nhất vào giới tính: Phụ nữ có tỷ lệ sống sót vượt trội (74.2%) so với nam giới (18.89%), thể hiện rõ quy tắc ưu tiên sơ tán là phụ nữ và trẻ em sẽ được ưu tiên hàng đầu
==> TỔng kết :Cơ hội sống sót cao nhất trên tàu Titanic thuộc về nữ giới, mua vé hạng 1 (giá cao) và đi cùng nhóm gia đình nhỏ 